### PR TITLE
gcp - add load balancing service initial setup

### PR DESCRIPTION
**Pull request by egor-miasnikov**
Added initial classes with `query` and `get` functionality for the following resources in `LoadBalancing` service

- Addresses
- Backend buckets
- Backend Sevices
- Forwarding Rules
- Global Addresses
- Global Forwarding Rules
- Health Checks
- HTTP Health Checks
- HTTPS Health Checks
- SSL Certificates
- SSL Policies
- Target HTTP Proxies
- Target HTTPS Proxies
- Target Instances
- Target Pools
- Target SSL Proxies
- Target TCP Proxies
- URL Maps

### CODE REVIEW

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from c7n_gcp.provider import resources
+from c7n_gcp.query import QueryResourceManager, TypeInfo
+
+
+@resources.register('loadbalancing-address')
+class LBAddress(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: I would change the classes to spell out LB to be consistent with the registry/class naming I've seen elsewhere.

e.g. `LoadBalancingAddress`

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'region': resource_info['region'],
+                'address': resource_info['name']})
+
+
+@resources.register('loadbalancing-url-map')
+class LBUrlMap(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'urlMap': resource_info['name']})
+
+
+@resources.register('loadbalancing-target-tcp-proxy')
+class LBTargetTcpProxy(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'targetTcpProxy': resource_info['name']})
+
+
+@resources.register('loadbalancing-target-ssl-proxy')
+class LBTargetSslProxy(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'targetSslProxy': resource_info['name']})
+
+
+@resources.register('loadbalancing-ssl-policy')
+class LBSslPolicy(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'sslPolicy': resource_info['name']})
+
+
+@resources.register('loadbalancing-ssl-certificate')
+class LBSslCertificate(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'sslCertificate': resource_info['name']})
+
+
+@resources.register('loadbalancing-target-https-proxy')
+class LBTargetHttpsProxy(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'targetHttpsProxy': resource_info['name']})
+
+
+@resources.register('loadbalancing-backend-bucket')
+class LBBackendBucket(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'backendBucket': resource_info['name']})
+
+
+@resources.register('loadbalancing-https-health-check')
+class LBHttpsHealthCheck(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'httpsHealthCheck': resource_info['name']})
+
+
+@resources.register('loadbalancing-http-health-check')
+class LBHttpHealthCheck(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'httpHealthCheck': resource_info['name']})
+
+
+@resources.register('loadbalancing-health-check')
+class LBHealthCheck(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'healthCheck': resource_info['name']})
+
+
+@resources.register('loadbalancing-target-http-proxy')
+class LBTargetHttpProxy(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'targetHttpProxy': resource_info['name']})
+
+
+@resources.register('loadbalancing-backend-service')
+class LBBackendService(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'backendService': resource_info['name']})
+
+
+@resources.register('loadbalancing-target-instance')
+class LBTargetInstance(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'zone': resource_info['zone'].rsplit('/', 1)[-1],
+                'targetInstance': resource_info['name']})
+
+
+@resources.register('loadbalancing-target-pool')
+class LBTargetPool(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'region': resource_info['region'].rsplit('/', 1)[-1],
+                'targetPool': resource_info['name']})
+
+
+@resources.register('loadbalancing-forwarding-rule')
+class LBForwardingRule(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'region': resource_info['region'].rsplit('/', 1)[-1],
+                'forwardingRule': resource_info['name']})
+
+
+@resources.register('loadbalancing-global-forwarding-rule')
+class LBGlobalForwardingRule(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_command('get', {
+                'project': resource_info['project_id'],
+                'forwardingRule': resource_info['name']})
+
+
+@resources.register('loadbalancing-global-address')
+class LBGlobalAddress(QueryResourceManager):
```
Comment (jondavidi) [inline tools/c7n_gcp/c7n_gcp/resources/loadbalancing.py]: see above comment

File: tools/c7n_gcp/tests/test_loadbalancing.py [hunk 19]
```diff
@@ -0,0 +1,530 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from gcp_common import BaseTest
+
+
+class LBAddressTest(BaseTest):
```
Comment (jondavidi) [inline tools/c7n_gcp/tests/test_loadbalancing.py]: please spell out LB in these class names as well

Comment (egor-miasnikov) [reply tools/c7n_gcp/tests/test_loadbalancing.py]: done

**Comment by egor-miasnikov** [comment]
@jondavidi I've fixed all comments.

**Comment by kapilt** [comment]
ideally this would be broken up into several prs, anything over 1k lines is frankly too big.

**Comment by egor-miasnikov** [comment]
> ideally this would be broken up into several prs, anything over 1k lines is frankly too big.

@kapilt maybe for now is just fine? Here is just a simple classes declaration with a test and without huge business logic?

**Comment by kapilt** [comment]
even though its mostly declarative I'd still recommend the break out, we've tried to set that as precedent so its a common guide, making exceptions invites more questions. Also there are several here that I don't see a particularly good use case  for, so invite additional scrutiny. flow is better small and constant, not icebergs ;-)

**Comment by bclements** [comment]
Please break up into smaller prs.

**Pull request closed without merging**